### PR TITLE
Allow updating of password on public share

### LINKED ISSUE #10671: Updating a password using OCS Share API changes the share ID
If you make an API call using the OCS Share API and set a password on a share, it appears to change the share ID. Not sure if this is intended or not.

Paging @PVince81 as requested.

### PR DESCRIPTION
**Pull request by rullzer**
PR to resolve #10671 and #14826 (and possibly others).

currently we cannot update passwords on public shares. What happens is that a new share is created. This leads to problems with the expiration date being not properly propagated and the share getting a new id. Several things have to happen:
- [x] Add setPassword function to share.php
- [x] Update OCS call
- [x] Update ajax/share.php call
- [x] Update share.js
- [x] Cleanup share.php code for links
- [ ] More JS unit test
- [ ] Fix issue with allow upload

### CODE REVIEW

File: lib/private/share/share.php [hunk 1]
```diff
@@ -1123,6 +1123,38 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+	 * @param string $itemType
+	 * @param string $itemSource
+	 * @param string $password
+	 * @throws \Exception
+	 * @return boolean
+	 */
+	public static function setPassword($itemType, $itemSource, $password) {
+		$user = \OC_User::getUser();
+
+		if ($password == '') {
```
Comment (LukasReschke) [inline lib/private/share/share.php]: `===` as per code guidelines

File: lib/private/share/share.php [hunk 2]
```diff
@@ -1123,6 +1123,38 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
 	/**
+	 * Set expiration date for a share
+	 * @param string $itemType
+	 * @param string $itemSource
+	 * @param string $password
+	 * @throws \Exception
+	 * @return boolean
+	 */
+	public static function setPassword($itemType, $itemSource, $password) {
+		$user = \OC_User::getUser();
```
Comment (LukasReschke) [inline lib/private/share/share.php]: `\OCP\IUserSession::getUser` with a check if the user is null and throwing an exception then?

File: lib/private/share/share.php [hunk 3]
```diff
@@ -1123,6 +1123,38 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+
+		if ($password == '') {
+			$password = null;
+		}
+
+		//If passwords are enforced the password can't be null
+		if (self::enforcePassword() && $password == null) {
```
Comment (LukasReschke) [inline lib/private/share/share.php]: `===` or `is_null `

File: lib/private/share/share.php [hunk 4]
```diff
@@ -1123,6 +1123,38 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
+			throw new \Exception('Cannot remove password');
+		}
+
+		$query = \OC_DB::prepare('UPDATE `*PREFIX*share` SET `share_with` = ? WHERE `item_type` = ? AND `item_source` = ?  AND `uid_owner` = ? AND `share_type` = ?');
```
Comment (LukasReschke) [inline lib/private/share/share.php]: Can you try to use the Doctrine Query Builder? – See https://github.com/owncloud/core/issues/12502

**Comment by rullzer** [comment]
@LukasReschke thanks for the feedback. Most stuff is just the result of copy pasting to get it to work. Cleaning up now.

File: lib/private/share/share.php [hunk 5]
```diff
@@ -1123,6 +1110,43 @@ public static function setExpirationDate($itemType, $itemSource, $date, $shareTi
...
 	/**
+	 * Set expiration date for a share
+	 * @param string $itemType
+	 * @param string $itemSource
+	 * @param string $password
+	 * @throws \Exception
+	 * @return boolean
+	 */
+	public static function setPassword($itemType, $itemSource, $password) {
+		$user = \OC::$server->getUserSession()->getUser();
```
Comment (LukasReschke) [inline lib/private/share/share.php]: For unit test purposes you might want to inject the UserSession as well as the DatabaseConnection in the functions signature.

(obviously not inject it in the public API as that would look ugly - but here it is a valid approach for static code :see_no_evil:)

**Comment by rullzer** [comment]
@owncloud-bot retest this please

File: lib/private/share/share.php (hunk 5 above)
Comment (rullzer) [inline lib/private/share/share.php]: yeah seems like a good intermediate solution... since rewriting share.php is clearly not in the scope of this PR

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by rullzer** [comment]
Please review @schiesbn @PVince81 @LukasReschke @DeepDiver1975

File: lib/public/idbconnection.php [hunk 6]
```diff
@@ -174,4 +174,11 @@ public function dropTable($table);
 	 * @return bool
 	 */
 	public function tableExists($table);
+
+	/**
+	 * Gets the QueryBuilder
+	 *
+	 * @return \Doctrine\DBAL\Query\QueryBuilder
+	 */
+	public function createQueryBuilder();
```
Comment (PVince81) [inline lib/public/idbconnection.php]: Interface change, @icewind1991 @DeepDiver1975 @schiesbn I guess we need this anyway ?

Comment (rullzer) [inline lib/public/idbconnection.php]: Good call. as on my todo list to ask more people about this. We somehow need it if we want to support the QueryBuilder or we can't use the IDBConnection to obtain it

**Comment by PVince81** [comment]
Nice one!

Originally I was hoping to have a more REST-style API where one must send the whole share model (all values), see https://github.com/owncloud/core/pull/11674

But I think for now we can live with this until we move everything to the OCS Share API.

**Comment by rullzer** [comment]
@PVince81 I agree. But for now this should be a workable solution. Changing the OCS API is something that probably needs to happen. But that is most likely a slow process.

**Comment by PVince81** [comment]
#### Tests
- [x] share with link, open link, change password: link still valid
- [ ] share with link, set expiration, allow upload: :no_entry_sign: FAIL it says "only 1 public share allowed"

**Comment by PVince81** [comment]
Ideally you should also add/update the JS unit tests, see here for an example: https://github.com/owncloud/core/pull/11674/files#diff-760d15f0463b1b3a6ca4b48223badbafR289

**Comment by rullzer** [comment]
@PVince81, the reason the second tests fails is because we try to create a new share. I did remove some code at that point. So it is now broken in a different way.

Just chaning the ajax call to call setPermissions does not work. Since then a shareWith field is requierd in the code. We cannot provide this since it is often not known.

The reason it works in the OCS Api is because there we just fecth all info from a share by the id. The hacks to get this to work properly probably are not pretty.

So I suggest to just move that call to the OCS API. Getting the javascript to act nice is probably way easier than hacking the ajax/share.php stuff (and will lead to much duplicated code). Like we already talked about. The new share modal should use more OCS stuff anyways.


On a related note, we should really start on cleaning up (or starting over) with lib/private/share/share.php :speak_no_evil:

**Comment by PVince81** [comment]
How about splitting the OCS Share fix from the ajax ones ? At least getting the OCS Share work has higher priority (especially if we intend to move the ajax call to use that too).

And in this way the risk of regression will hopefully be lower. What do you think ?

**Comment by rullzer** [comment]
Yeah that sounds probably best.
I'll have a look later.
You guys can then have a look at the shiny new model with OCS stuff next week :wink:

**Comment by rullzer** [comment]
See #14987 for a PR only for the OCS API. Closing this one for now.

**Pull request closed without merging**